File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {

    List<List<Integer>> lists = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<Integer> listOfNumbers = new ArrayList<>();
        dfs(nums,0,listOfNumbers,target,0);
        return lists;
    }

    public void dfs(int[] nums, int index, 
        List<Integer> listOfNumbers, 
        int target, int currentSum){
        
        if(currentSum == target){
            lists.add(new ArrayList<>(listOfNumbers));
            return;
        }

        if(currentSum > target || index >= nums.length){
            return;
        }

        listOfNumbers.add(nums[index]);
        currentSum+=nums[index];
        
        dfs(nums, index, listOfNumbers, target, currentSum);

        listOfNumbers.remove(listOfNumbers.size() - 1);
        currentSum-=nums[index];
        
        dfs(nums, index + 1, listOfNumbers, target, currentSum);
    }
}

File: Data Structures & Algorithms/missing-element-in-sorted-array/submission-0.py
class Solution:
    def missingElement(self, nums: List[int], k: int) -> int:
        n = len(nums)
        l,r = 0, n-1

        while l < r:
            mid = r-(r-l) // 2
            if (nums[mid] - nums[0]) - mid < k:
                l = mid
            else:
                r = mid -1
        
        return nums[0] + k + l

        